File: LinkedList/DeepCopy-LinkedList/solution_recursive.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

# Dictionary which hold old nodes as keys and new nodes as its values. 
visitedDict = {}
class Solution(object):
    """
    :type head: RandomListNode
    :rtype: RandomListNode
    """

    def copyRandomList(self, head):
        global visitedDict
        if head == None:
            return None
        
        # If there is value in the dictionary corresponding to the node to be copied
        # then return the respective value from the visited dictionary
        if head in visitedDict:
            return visitedDict[head]

        # create a new node
        # with the label same as old node. 
        node = RandomListNode(head.label)

        # save this value in the dictionary, this is needed since there might be dependency loops due 
        # to randomness of random pointers.
        # this would help in having a object reference saved in the dictionary already
        # which could be updated later.
        visitedDict[head] = node

        # creating a copy of next and random nodes by calling the function recursively
        node.next = self.copyRandomList(head.next)
        node.random = self.copyRandomList(head.random)

        # update the new node pointer to save the updated references.
        visitedDict[head] = node
        
        return node

File: Linked-List/DeepCopy-LinkedList/solution_recursive.java
public class Solution {
  // HashMap which holds old nodes as keys and new nodes as its values.
  HashMap<RandomListNode, RandomListNode> visitedHash =
          new HashMap<RandomListNode, RandomListNode>();

  public RandomListNode copyRandomList(RandomListNode head) {

    if (head == null) {
      return null;
    }

    // If we have already processed the current node, then we simply return the cloned version of it.
    if (this.visitedHash.containsKey(head)) {
      return this.visitedHash.get(head);
    }

    // Create a new node with the label same as old node. (i.e. copy the node)
    RandomListNode node = new RandomListNode(head.label);

    // Save this value in the hash map. This is needed since there might be
    // loops during traversal due to randomness of random pointers and this would help us avoid them.
    this.visitedHash.put(head, node);

    // Recursively copy the remaining linked list starting once from the next pointer and then from the random pointer.
    // Thus we have two independent recursive calls.
    // Finally we update the next and random pointers for the new node created.
    node.next = this.copyRandomList(head.next);
    node.random = this.copyRandomList(head.random);

    return node;
  }
}